File: js/src/policy.ts
import type { Decision, Policy, Resolution } from "./types.js";

const DEFAULT_STATUSES = ["active"];

/**
 * Maps a resolution + policy to a decision.
 *
 * - `allow` — registered, active, and meets the identity bar.
 * - `deny`  — present but fails the policy (wrong status / unverified when required),
 *             or absent/failed-lookup under a strict policy.
 * - `flag`  — present-but-unknown or fail-open outage; serve but tag for logging.
 *
 * Fails open by default: when the lookup could not complete (`ok === false`),
 * returns `flag` unless `policy.failClosed` is set.
 */
export function decide(resolution: Resolution, policy: Policy = {}): Decision {
  const allowStatuses = policy.allowStatuses ?? DEFAULT_STATUSES;

  if (!resolution.ok) return policy.failClosed ? "deny" : "flag";

  if (!resolution.found) return policy.requireVerified ? "deny" : "flag";

  if (resolution.status && !allowStatuses.includes(resolution.status))
    return "deny";

  if (policy.requireVerified && resolution.identityLevel !== "verified")
    return "deny";

  return "allow";
}

File: python/src/wayid_verify/policy.py
from .types import Decision, Policy, Resolution


def decide(resolution: Resolution, policy: Policy = None) -> Decision:
    """Map a resolution + policy to ``allow`` / ``deny`` / ``flag``.

    Fails open: when the lookup could not complete (``ok is False``), returns
    ``flag`` unless ``policy.fail_closed`` is set.
    """
    policy = policy or Policy()

    if not resolution.ok:
        return "deny" if policy.fail_closed else "flag"

    if not resolution.found:
        return "deny" if policy.require_verified else "flag"

    if resolution.status and resolution.status not in policy.allow_statuses:
        return "deny"

    if policy.require_verified and resolution.identity_level != "verified":
        return "deny"

    return "allow"
